# Citation Faithfulness: When NLI Lies About RAG Answers

**Andrea Trugenberger** · SCIPER 357615 · CS-552 Spring 2026 ·
Team Faithful RAG · M2 Progress

---

## Question

When a language model gives an answer with a citation, how do you actually
check that the cited passage supports the claim? Our proposal said to use
NLI (DeBERTa-v3), which is the standard approach in the literature. But
the more I worked with it, the more I started wondering:

**Is NLI actually good enough on its own, or does it miss claims that
sound like they belong but are factually wrong?**

Short version: it's not good enough. When I ran a closed-book test on
three LLMs, NLI marked about 75% of the made-up answers as faithful. An
LLM judge caught them. The gap was 55 to 61 percentage points and showed
up across every model.

## What I built

* **`evaluation/faithfulness/faithfulness_scorer.py`**, the verification
  pipeline. It has four functions: `extract_claims` (splits an answer
  into atomic claims, FActScore-style), `verify_claim_nli` (zero-shot
  DeBERTa), `verify_claim_llm_judge` (GPT-4o-mini with a prompt asking
  it to check whether the specific entities and terms in the claim are
  actually in the passage), and `verify_claim` (the orchestrator that
  picks between `NO_CITATION`, `FABRICATED_SOURCE`, `SUPPORTED`, and
  `NOT_SUPPORTED`).
* **`scripts/run_faithfulness_experiment_full.py`**, the script I used
  to run the experiment. It loads the 24 answerable questions from
  `gold_qa.json`, asks each of three LLMs to answer closed-book, then
  runs both verifiers on every claim that comes back.
* **Two write-ups**: `01_llm_judge_vs_nli.md` (the pilot I did first on
  one question) and `02_full_results.json` (the raw output of the full
  run that this notebook reads).

## Design choice: four labels, not just yes/no

At first I thought of faithfulness as a yes/no question: does the cited
passage back up the claim, or not? But while writing the code I realized
there are actually four different things that can go wrong. Sometimes
the model makes a claim without citing anything at all (`NO_CITATION`).
Sometimes it cites a paper that isn't in our corpus, basically inventing
a source (`FABRICATED_SOURCE`). And sometimes the citation is real but
the passage either backs up the claim (`SUPPORTED`) or doesn't
(`NOT_SUPPORTED`). The first two cases are easy to spot without running
any model, since you just check if the citation exists. Only the last
two actually need a verifier. Separating them made the code a lot
cleaner.

## Design choice: two verifiers, not one

NLI was the obvious starting point because that's what ALCE and
FActScore use. But pretty early on I tried this example: a passage
that says *"Shapiro identifies two transaction costs: complements and
hold-up"* paired with a fake claim that says *"Shapiro identifies three
transaction costs including a licensing problem."* The claim is wrong
on every specific detail. NLI labelled it `supported` with confidence
0.984, literally the same confidence it gave to a correct claim.

What I realized is that NLI is basically checking whether the two texts
are about the same general topic. Both the passage and the claim talk
about Shapiro and transaction costs, so there's enough overlap for NLI
to call it entailment. It just doesn't look closely enough at the
specifics to notice they don't match. That's why I added the LLM judge
as a second check. The two verifiers do different things: NLI is fast
and free but blind to specifics, while the LLM judge is slower and
costs API calls but reads the text more carefully.

## Experimental setup

For each of the 24 answerable gold questions, I asked three LLMs
(GPT-4o-mini, Claude 3.5 Haiku, DeepSeek) to answer closed book
meaning no retrieval, just the question. The point was to force the
model to hallucinate from training data so the verifiers would actually
have something to flag. Then for each answer, I split it into atomic
claims and ran both verifiers against the gold passage from
`gold_qa.json`.

Closed-book is the hardest setting for the model and the easiest one
to see the verifier gap in. With retrieval the gap would shrink because
the LLM would have the right text in its prompt. I figured the
closed-book version was the cleanest way to show the failure mode.



In [1]:
"""Defensive install: make sure pandas is available on whatever environment the
notebook is opened in (cluster pod, fresh venv, etc.). Idempotent on rerun."""
import subprocess, sys
try:
    import pandas  # noqa: F401
except ImportError:
    subprocess.check_call([sys.executable, "-m", "pip", "install", "--quiet", "pandas"])
    import pandas  # noqa: F401
print(f"pandas {pandas.__version__} ready")


pandas 2.3.3 ready


In [3]:
"""Set up paths so we can load the experiment results."""
import json
import sys
from pathlib import Path

repo_root = Path.cwd().resolve()
while not (repo_root / "pyproject.toml").exists():
    if repo_root.parent == repo_root:
        raise RuntimeError("Repo root not found.")
    repo_root = repo_root.parent
if str(repo_root) not in sys.path:
    sys.path.insert(0, str(repo_root))

results_path = repo_root / "evaluation/faithfulness/results/02_full_results.json"
all_results = json.loads(results_path.read_text())

n_runs = len(all_results)
models = sorted({r["model"] for r in all_results if "model" in r})
print(f"Total (question, model) runs: {n_runs}")
print(f"Models compared: {models}")
print(f"Questions covered: {len({r['question_id'] for r in all_results})}")


Total (question, model) runs: 72
Models compared: ['anthropic/claude-3.5-haiku', 'deepseek/deepseek-chat', 'openai/gpt-4o-mini']
Questions covered: 24


## Headline numbers

In [4]:
"""Aggregate per-model faithfulness across all questions."""
import pandas as pd

rows = []
for model in models:
    model_rows = [r for r in all_results if r.get("model") == model and "error" not in r]
    n = len(model_rows)
    nli_avg = sum(r["nli_faithfulness"] for r in model_rows) / n
    judge_avg = sum(r["judge_faithfulness"] for r in model_rows) / n
    rows.append({
        "model": model,
        "n_questions": n,
        "total_claims": sum(r["n_claims"] for r in model_rows),
        "NLI faithfulness": round(nli_avg, 3),
        "LLM-judge faithfulness": round(judge_avg, 3),
        "gap (NLI - judge)": round(nli_avg - judge_avg, 3),
    })

headline_df = pd.DataFrame(rows).set_index("model")
headline_df


,n_questions,total_claims,NLI faithfulness,LLM-judge faithfulness,gap (NLI - judge)
model,,,,,
anthropic/claude-3.5-haiku,24,174,0.75,0.173,0.577
deepseek/deepseek-chat,24,178,0.75,0.198,0.552
openai/gpt-4o-mini,24,183,0.75,0.138,0.612


**Reading the table.**

* **All three generators get 75.0% NLI faithfulness.** NLI rubber-stamps anything topically related as `supported`. It cannot tell the three models apart because it cannot see what each got specifically wrong.
* **The LLM judge separates them.** GPT-4o-mini 13.8%, Claude 3.5 Haiku 17.3%, DeepSeek-chat 19.8%.
* **The gap is 55–61 percentage points and is stable across providers.** This is the headline finding: a systematic failure of NLI as a citation-faithfulness gate.


In [5]:
"""Case study: q004 — the two transaction costs in Shapiro 2001."""
q004 = [r for r in all_results if r.get("question_id") == "q004"]
for r in q004:
    if "error" in r:
        continue
    print(f"\n=== {r['model']} ===")
    print(f"Answer: {r['answer'][:300]}...")
    print(f"NLI:   {r['nli_faithfulness']:.1%}    Judge: {r['judge_faithfulness']:.1%}")
    print("Per-claim (first 3):")
    for c in r["claims"][:3]:
        print(f"  NLI={c['nli_label']:<14}  Judge={c['judge_label']:<14}  {c['claim'][:80]}")



=== openai/gpt-4o-mini ===
Answer: According to Shapiro (2001), the two main types of transaction costs imposed by overlapping patents on innovation are negotiation costs and enforcement costs. Negotiation costs arise from the need to obtain licenses from multiple patent holders, while enforcement costs relate to the challenges and e...
NLI:   100.0%    Judge: 0.0%
Per-claim (first 3):
  NLI=supported       Judge=not_supported   According to Shapiro (2001), there are two main types of transaction costs impos
  NLI=supported       Judge=not_supported   The two main types of transaction costs are negotiation costs and enforcement co
  NLI=supported       Judge=not_supported   Negotiation costs arise from the need to obtain licenses from multiple patent ho

=== anthropic/claude-3.5-haiku ===
Answer: According to Shapiro (2001), the two main types of transaction costs imposed by overlapping patents are: (1) search costs, which arise from the difficulty and expense of identifying and locat

**What happened on q004.** Gold answer: complements problem and hold-up problem. None of the three models produced these names. GPT-4o-mini said "negotiation costs" and "enforcement costs"; Claude said "search costs" and "licensing costs"; DeepSeek said "search costs" and "bargaining costs". NLI labelled all of them supported. The judge correctly labelled them not-supported.

## q017: invented numbers


In [6]:
"""Case study: q017 — annual US patent issuance growth 1983-2010."""
q017 = [r for r in all_results if r.get("question_id") == "q017"]
for r in q017:
    if "error" in r:
        continue
    print(f"\n=== {r['model']} ===")
    print(f"Answer: {r['answer'][:300]}...")
    print(f"NLI:   {r['nli_faithfulness']:.1%}    Judge: {r['judge_faithfulness']:.1%}")



=== openai/gpt-4o-mini ===
Answer: According to Boldrin and Levine (2013), annual US patent issuance grew from approximately 90,000 patents in 1983 to about 200,000 patents in 2010, indicating a significant increase over that period. This growth reflects a more than twofold increase in the number of patents issued annually....
NLI:   100.0%    Judge: 25.0%

=== anthropic/claude-3.5-haiku ===
Answer: According to Boldrin and Levine (2013), annual US patent issuance grew from approximately 50,000 patents per year in 1983 to around 250,000 patents per year in 2010, representing a five-fold increase over this 27-year period. This substantial growth in patent issuance occurred across various technol...
NLI:   100.0%    Judge: 14.3%

=== deepseek/deepseek-chat ===
Answer: According to Boldrin and Levine (2013), annual U.S. patent issuance grew significantly between 1983 and 2010, increasing from approximately 62,000 patents to over 240,000 patents. This represents nearly a fourfold growth o

**What happened on q017.** Three LLMs, three different made-up numbers for the same factual question. NLI gave all three 100% because the topic matches. The judge correctly catches the specific number mismatches.

## q018: a moment of honesty


In [7]:
"""Case study: q018 — Claude's honest refusal."""
q018 = [r for r in all_results if r.get("question_id") == "q018"]
for r in q018:
    if "error" in r or "claude" not in r["model"]:
        continue
    print(f"=== {r['model']} ===")
    print(f"Answer: {r['answer']}")
    print(f"NLI:   {r['nli_faithfulness']:.1%}    Judge: {r['judge_faithfulness']:.1%}")


=== anthropic/claude-3.5-haiku ===
Answer: I apologize, but I cannot confidently answer this question without having direct access to Moser's 2013 research. Without reading the specific paper, I cannot accurately represent the author's conclusions about patent laws and innovation. To provide a reliable response, I would need to review the original source material.
NLI:   100.0%    Judge: 0.0%


**What happened on q018.** Claude said *"I cannot confidently answer this without direct access to Moser's paper."* The right answer for a closed-book question. GPT and DeepSeek both hallucinated specifics on the same question.

## Limitations

1. **Closed-book is the worst case for the generators.** With retrieval the gap would shrink. The 55–61 pp gap is an upper bound on verifier disagreement.
2. **The gold dataset has no `not_supports` claims.** Every gold claim is true with a real supporting passage. The experiment measures the verifier gap on potentially hallucinated generated claims, but cannot measure verifier precision on hand-crafted false claims. Faruk's adversarial controls (q900–q905) are starting to fix this.
3. **The `FABRICATED_SOURCE` path is unused here.** The orchestrator handles this case (tested separately on a fake paper id), but empirical evaluation is M3 work.

## M3 plan

* Re-run the cross-LLM experiment **with retrieval** using Elie's best retriever (E5-large coarse rerank).
* Run both verifiers on the **adversarial control set** to measure precision on intentionally false claims.
* Try the larger NLI model originally specified in the proposal to see how much of the gap is small-model capacity versus NLI as a technique.
